govcd/cse.go

```diff
@@ -0,0 +1,392 @@
+package govcd
+
+import (
+	"encoding/json"
+	"fmt"
+	semver "github.com/hashicorp/go-version"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+	"github.com/vmware/go-vcloud-director/v2/util"
+	"strings"
+	"time"
+)
+
+// CseCreateKubernetesCluster creates a Kubernetes cluster with the data given as input (CseClusterSettings). If the given
+// timeout is 0, it waits forever for the cluster creation.
+//
+// If the timeout is reached and the cluster is not available (in "provisioned" state), it will return a non-nil CseKubernetesCluster
+// with only the cluster ID and an error. This means that the cluster will be left in VCD in any state, and it can be retrieved afterward
+// with Org.CseGetKubernetesClusterById and the returned ID.
+//
+// If the cluster is created correctly, returns all the available data in CseKubernetesCluster or an error if some of the fields
+// of the created cluster cannot be calculated or retrieved.
+func (org *Org) CseCreateKubernetesCluster(clusterData CseClusterSettings, timeout time.Duration) (*CseKubernetesCluster, error) {
+	clusterId, err := org.CseCreateKubernetesClusterAsync(clusterData)
+	if err != nil {
+		return nil, err
+	}
+
+	err = waitUntilClusterIsProvisioned(org.client, clusterId, timeout)
+	if err != nil {
+		return &CseKubernetesCluster{
+			client: org.client,
+			ID:     clusterId,
+		}, err
+	}
+
+	return getCseKubernetesClusterById(org.client, clusterId)
+}
+
+// CseCreateKubernetesClusterAsync creates a Kubernetes cluster with the data given as input (CseClusterSettings), but does not
+// wait for the creation process to finish, so it doesn't monitor for any errors during the process. It returns just the ID of
+// the created cluster. One can manually check the status of the cluster with VCDClient.CseGetKubernetesClusterById and the result of this method.
+func (org *Org) CseCreateKubernetesClusterAsync(clusterSettings CseClusterSettings) (string, error) {
+	if org == nil {
+		return "", fmt.Errorf("CseCreateKubernetesClusterAsync cannot be called on a nil Organization receiver")
+	}
+
+	cseSubcomponents, err := getCseComponentsVersions(clusterSettings.CseVersion)
+	if err != nil {
+		return "", err
+	}
+
+	internalSettings, err := clusterSettings.toCseClusterSettingsInternal(*org)
+	if err != nil {
+		return "", fmt.Errorf("error creating the CSE Kubernetes cluster: %s", err)
+	}
+
+	payload, err := internalSettings.getUnmarshaledRdePayload()
+	if err != nil {
+		return "", err
+	}
+
+	rde, err := createRdeAndGetFromTask(org.client, cseKubernetesClusterVendor, cseKubernetesClusterNamespace, cseSubcomponents.CapvcdRdeTypeVersion,
+		types.DefinedEntity{
+			EntityType: internalSettings.RdeType.ID,
+			Name:       internalSettings.Name,
+			Entity:     payload,
+		}, &TenantContext{
```

Comment:
No need, but there is `org.getTenantContext()` :)

Reply:
Changed, more elegant now.